src/main/java/org/truffleruby/core/array/ArrayWriteNormalizedNode.java

```diff
@@ -41,16 +41,49 @@ public Object writeWithin(DynamicObject array, int index, Object value,
         return value;
     }
 
+    // writing within an array after converting it to be mutable
+
+    @Specialization(guards = { "!strategy.isStorageMutable()",
+            "isInBounds(array, index)", "strategy.matches(array)", "mutableStrategy.accepts(value)"
+    }, limit = "STORAGE_STRATEGIES")
+    public Object writeWithin(DynamicObject array, int index, Object value,
+            @Cached("of(array)") ArrayStrategy strategy,
+            @Cached("strategy.generalizeForMutation()") ArrayStrategy mutableStrategy) {
+
+        final ArrayMirror oldStore = strategy.newMirror(array);
+        final ArrayMirror newStore = mutableStrategy.newArray(oldStore.getLength());
+        oldStore.copyTo(newStore, 0, 0, strategy.getSize(array));
+        strategy.setStore(array, newStore);
```

Comment:
Could have a helper method for this? It's quite verbose just to make a copy.
Maybe `ArrayStrategy#makeStorageMutable(array) => ArrayMirror` or so.

Reply:
That seems reasonable. I think also adding a method to `ArrayMirror` to extract a range and do possibly change array storage to be immutable would help too.